### PR TITLE
 More robust reference counting for streams. 

### PR DESCRIPTION
**Pull request by ryzhyk**
Only the last commit is new; the other one is PR #43

### CODE REVIEW

**Review by mbudiu-vmw — APPROVED** [review]

File: src/operator/z1.rs [hunk 1]
```diff
@@ -110,47 +110,211 @@ where
...
+        Self {
+            zero,
+            timestamp: 0,
+            val: vec![],
+        }
+    }
+
+    fn reset(&mut self) {
+        self.timestamp = 0;
+        self.val = vec![];
```
Comment (Kixiron) [inline src/operator/z1.rs:158]: Keep this buffer allocated
```suggestion
        self.val.clear();
```

File: src/operator/z1.rs [hunk 2]
```diff
@@ -110,47 +110,211 @@ where
...
+
+impl<T> Z1Nested<T>
+where
+    T: Clone,
+{
+    pub fn new(zero: T) -> Self {
+        Self {
+            zero,
+            timestamp: 0,
+            val: vec![],
```
Comment (Kixiron) [inline src/operator/z1.rs:152]:
```suggestion
            val: Vec::new(),
```

File: src/operator/z1.rs [hunk 3]
```diff
@@ -110,47 +110,211 @@ where
...
+
+    fn clock_start(&mut self, scope: Scope) {
+        if scope == 0 {
+            self.val.truncate(self.timestamp);
+        }
+        self.timestamp = 0;
+    }
+
+    fn clock_end(&mut self, scope: Scope) {
+        if scope > 0 {
```
Comment (Kixiron) [inline src/operator/z1.rs:178]:
```suggestion
        if scope != 0 {
```

File: src/operator/z1.rs [hunk 4]
```diff
@@ -110,47 +110,211 @@ where
...
+{
+    fn get_output(&mut self) -> T {
+        if self.timestamp >= self.val.len() {
+            assert_eq!(self.timestamp, self.val.len());
+            self.val.push(self.zero.clone());
+        }
+
+        let mut zero = self.zero.clone();
+        swap(
+            unsafe { self.val.get_unchecked_mut(self.timestamp) },
```
Comment (Kixiron) [inline src/operator/z1.rs:196]: Needs safety comment

File: src/operator/integrate.rs [hunk 5]
```diff
@@ -186,4 +270,65 @@ mod test {
             root.step().unwrap();
         }
     }
+
+    #[rustfmt::skip]
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/integrate.rs:274]: It's actually for the comment.

File: src/operator/z1.rs (hunk 1 above)
Comment (ryzhyk) [reply src/operator/z1.rs:158]: This is reset for the outer timestamp, basically the whole parent circuit starting from scratch, so there's no reason to expect that we'll need similar amount of inner iterations.  In any case, savings are negligible across many parent clock ticks.

File: src/operator/integrate.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kixiron) [reply src/operator/integrate.rs:274]: What do you mean, what does rustfmt do to it? Have you tried wrapping it in a code block instead?

[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/integrate.rs:274]: Good point, I was not thinking.

File: src/operator/z1.rs (hunk 1 above)
Comment (Kixiron) [reply src/operator/z1.rs:158]: I mean, there's no reason not to keep the memory allocated, it's only a benefit really

File: src/circuit/circuit_builder.rs [hunk 6]
```diff
@@ -55,6 +55,39 @@ use crate::circuit::{
...
 };
 
+/// Value stored in the stream.
+struct StreamValue<D> {
+    /// Value written to the stream at the current clock cycle;
+    /// `None` after the last consumer has retrieved the value from the stream.
+    val: Option<D>,
+    /// The number of consumers connected to the stream.  Each consumer
+    /// reads from the stream exactly once at every clock cycle.
+    readers: usize,
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/circuit_builder.rs:65]: `readers` -> `consumers` is nice. I think `tokens` is adequate though.

**Pull request merged**